### PR TITLE
fetch: implement new spec changes

### PR DESCRIPTION
**Pull request by KhafraDev**
Implements
1. https://github.com/whatwg/fetch/commit/813788a7d7b4bfba9d4398efd22fdae13125ee8a
2. https://github.com/whatwg/fetch/commit/4cd70cf5aecf777a9f6d451c6381423153913da9
3. https://github.com/whatwg/fetch/commit/4a7bf3536c4bd4cd30aedba5d0a2ba865c587699

### CODE REVIEW

File: lib/fetch/index.js [hunk 1]
```diff
@@ -850,13 +858,13 @@ async function schemeFetch (fetchParams) {
...
       //    header list is « (`Content-Type`, mimeType) »,
-      //    and body is dataURLStruct’s body.
+      //    and body is dataURLStruct’s body as a body.
       return makeResponse({
         statusText: 'OK',
         headersList: [
           ['content-type', mimeType]
         ],
-        body: extractBody(dataURLStruct.body)[0]
+        body: safelyExtractBody(dataURLStruct.body)[0]
```
Comment (ronag) [inline lib/fetch/index.js:867]: Why safely?

**Review by ronag — APPROVED** [review]

File: lib/fetch/body.js [hunk 2]
```diff
@@ -19,33 +19,45 @@ const { parseMIMEType, serializeAMimeType } = require('./dataURL')
...
+  } else if (isBlobLike(object)) {
+    // 3. Otherwise, if object is a Blob object, set stream to the
+    //    result of running object’s get stream.
+    stream = object.stream()
+  }
+
+  // 4. Otherwise, set stream to a new ReadableStream object, and set
+  //    up stream.
+
+  // 5. Assert: stream is a ReadableStream object.
```
[the remark this answers is not in the document]
Comment (KhafraDev) [reply lib/fetch/body.js:43]: See my note below:
```
// TODO(@KhafraDev): setting stream to an empty stream here causes
  // the process to unexpectedly exit. An error is likely being swallowed.
```
for some reason tap starts failing when I implement 4 & 5. No idea why 😕

File: lib/fetch/index.js (hunk 1 above)
Comment (KhafraDev) [reply lib/fetch/index.js:867]: https://fetch.spec.whatwg.org/#byte-sequence-as-a-body

"To get a [byte sequence](https://infra.spec.whatwg.org/#byte-sequence) bytes as a body, return the [body](https://fetch.spec.whatwg.org/#body-with-type-body) of the result of [safely extracting](https://fetch.spec.whatwg.org/#bodyinit-safely-extract) bytes."

File: lib/fetch/body.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (KhafraDev) [reply lib/fetch/body.js:43]: Seems like a node core bug...

```js
const readable = new ReadableStream({
  start () {},
  type: undefined
})

const reader = readable.getReader()

while (true) {
  const { done, value } = await reader.read()

  console.log({ done, value })

  if (done) {
    break
  }
}

console.log('hi')
```

'hi' is never logged.

File: lib/fetch/body.js [hunk 3]
```diff
@@ -19,33 +19,65 @@ const { parseMIMEType, serializeAMimeType } = require('./dataURL')
...
+    stream = object.stream()
+  } else {
+    // 4. Otherwise, set stream to a new ReadableStream object, and set
+    //    up stream.
+    stream = new ReadableStream({
+      async pull (controller) {
+        controller.enqueue(
+          typeof source === 'string' ? new TextEncoder().encode(source) : source
+        )
+        queueMicrotask(() => {
```
Comment (ronag) [inline lib/fetch/body.js:46]: Yes, this is a known issue with node readable streams. I think we have the same queueMicrotask hack somewhere else.

Comment (ronag) [reply lib/fetch/body.js:46]: Maybe we could move this into a `closeReadableStream` helper function and re-use? Then we can link additional information why that is needed in one place.

Comment (KhafraDev) [reply lib/fetch/body.js:46]: Yeah I see it somewhere else, I actually copied it from there lol

Do you know why it happens?

Comment (ronag) [reply lib/fetch/body.js:46]: I don't remember. Maybe @jasnell has an idea. I believe there is some issue over at the node repo.

Comment (KhafraDev) [reply lib/fetch/body.js]: added the helper

**Review by ronag — APPROVED** [review]

**Pull request merged**